src/ai/backend/agent/stage/kernel_lifecycle/docker/scratch.py

```diff
@@ -0,0 +1,320 @@
+import asyncio
+import functools
+import os
+import shutil
+import subprocess
+import sys
+from collections.abc import Iterable
+from dataclasses import dataclass
+from pathlib import Path
+from subprocess import CalledProcessError
+from typing import Optional, override
+
+import pkg_resources
+
+from ai.backend.common.docker import KernelFeatures
+from ai.backend.common.stage.types import Provisioner, ProvisionStage, SpecGenerator
+from ai.backend.common.types import BinarySize, KernelId
+
+
+@dataclass
+class ContainerOwnershipConfig:
+    kernel_uid: Optional[int]
+    kernel_gid: Optional[int]
+    supplementary_gids: set[int]
+
+    fallback_kernel_uid: int
+    fallback_kernel_gid: int
+
+    kernel_features: frozenset[str]
+
+
+@dataclass
+class ScratchSpec:
+    kernel_id: KernelId
+
+    container_config: ContainerOwnershipConfig
+
+    scratch_type: str
+    scratch_root: Path
+    scratch_size: BinarySize
+
+
+class ScratchSpecGenerator(SpecGenerator[ScratchSpec]):
+    @override
+    async def wait_for_spec(self) -> ScratchSpec:
+        """
+        Waits for the spec to be ready.
+        """
+        # In a real implementation, this would wait for some condition to be met.
+        return ScratchSpec()
+
+
+@dataclass
+class ScratchResult:
+    scratch_root: Path
+    scratch_type: str
+    kernel_id: KernelId
```

Comment:
It seems a bit unnatural to include the `kernel_id` as a result value.
If another stage needs the `kernel_id`, it might appear as though there's a dependency on scratch provision stage, and the dependency relationship wouldn't be easily visible.